lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+#[macro_use]
+extern crate lazy_static;
+
+use libc::{fd_set, select, FD_SET};
+use rdp::core::client::{Connector, RdpClient};
+use rdp::core::event::*;
+use rdp::model::error::*;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::mem;
+use std::net::TcpStream;
+use std::os::unix::io::AsRawFd;
+use std::ptr;
+use std::sync::{Arc, Mutex};
+
+struct Client {
+    rdp_client: RdpClient<TcpStream>,
+    tcp_fd: usize,
+}
+type SyncRdpClient = Arc<Mutex<Client>>;
+
+// Rust-side registry of clients, to allow Go to reference a specific client in calls after
+// connect_rdp.
+lazy_static! {
+    static ref RDP_CLIENTS: Arc<Mutex<HashMap<i64, SyncRdpClient>>> =
+        Arc::new(Mutex::new(HashMap::new()));
+}
+
+fn register_client(client_ref: i64, client: Client) {
+    RDP_CLIENTS
+        .lock()
+        .unwrap()
+        .insert(client_ref, Arc::new(Mutex::new(client)));
+}
+
+fn unregister_client(client_ref: &i64) {
+    RDP_CLIENTS.lock().unwrap().remove(client_ref);
+}
+
+#[derive(Debug)]
+enum ClientError {
+    RDP(Error),
+    ClientNotFound,
+    Other(String),
+}
+
+impl From<Error> for ClientError {
+    fn from(e: Error) -> ClientError {
+        ClientError::RDP(e)
+    }
+}
+
+fn with_client<F: FnMut(&SyncRdpClient) -> Result<(), ClientError>>(
```

Comment:
Usually `std::option::Option` is preferred to semantically encode the notion of something potentially not existing while `std::result::Result` is used to encode the notion of an operation expected to succeed going wrong. This to me seems like the former.